File: javascript/ex9.js
// using a dictionary to map the alphabetised columns to numbers
let columns = {"a": 1, "b": 2, "c": 3, "d": 4, "e": 5, "f": 6, "g": 7, "h": 8};

// using a array of arrsys to represent the 8 possible moves a Knight can make
// e.g. first tuple = along one column, up two rows
let posMoves = [[1, 2], [1, -2], [2, 1], [2, -1], [-1, 2], [-1, -2], [-2, 1], [-2, -1]];

function ex9(start, end, moves){
    // transform the input to [column, row] format as nums in list
    let position = [columns[start[0]], parseInt(start[1])];
    let endPosition = [columns[end[0]], parseInt(end[1])];

    console.log(position);
    console.log(endPosition);

    // create a dictionary where key = move_num and value = list of pos_positions starting move 0 = position
    let posPositions = {0: [position]}

    // limit the number of moves
    for (let num = 0; num < moves; num++){
        let moveNum = num + 1;  // start from 1 (not 0)
        posPositions[moveNum] = [];  // create a new key in the dictionary for this move

        // iterate through the positions after each move
        for (let pos = 0; pos < posPositions[num].length; pos++) {

            // adding the possible moves to the current position
            for (let move = 0; move < posMoves.length; move++) {
                let newPosition = [posPositions[num][pos][0] + posMoves[move][0], posPositions[num][pos][1] + posMoves[move][1]];
                // is it a match for the end point?   
                if (newPosition[0] == endPosition[0] && newPosition[1] == endPosition[1]) {
                    console.log("new position" + newPosition);
                    console.log(posPositions);
                    return true;
                }
                // add new position to the dictionary if it is not out of bounds
                else if (0 < newPosition[0] && newPosition[0] <= 8 
                    && 0 < newPosition[1] && newPosition[1] <= 8)
                       {
                        posPositions[moveNum].push(newPosition);
                      }
                }        
        }
    }
    
    return false;
}
    
console.log(ex9("a1", "f5", 2));
console.log(ex9("a1", "f5", 3));
console.log(ex9("b2", "c3", 1));
console.log(ex9("a1", "c3", 5));
// console.log(ex9("a1", "f5", 2));
// console.log(ex9("a1", "f5", 2));

File: python/ex9.py
# write a function that takes the start and end point on a chess board and a number of moves a knoght can make
# start and end will be in "a1", "f3" format so cut these into columns and rows and change the columns to numbers
# if using recursion, the base case will be 0 when the knight has no movement left
# create a function that runs all 8 possibilities for the knight's movement and if the start == end return true
# if the knight goes out of bounds break this function

columns = { "a": 1, "b": 2, "c": 3, "d": 4, "e": 5, "f": 6, "g": 7, "h": 8}
pos_moves = [[1, 2], [1, -2], [2, 1], [2, -1], [-1, 2], [-1, -2], [-2, 1], [-2, -1]]

def ex9(start, end, moves):
    start_column = columns[start[0]]
    start_row = int(start[-1])

    start_position = [start_column, start_row]
    end_position = [columns[end[0]], int(end[-1])]

    #creat a list of possible positions (might need to make one per turn)
    pos_positions = [start_position]

    #limit the number of moves
    for num in range(moves, 0, -1):
        if num == 0:
            return False
        for position in pos_positions:
            #adding the possible moves to the start_position
            for move in pos_moves:
                new_position = [position[0] + move[0], position[1] + move[1]]
                if new_position == end_position:
                    return True
                elif 0 < new_position[0] and new_position[1] <= 8:
                    pos_positions.append(new_position)



    # print(start_position)
    # start_position += 1, 1
    #
    # return start_position

    # end_column = columns[end[0]]
    # end_row = end[-1]
    #
    # if start_column == end_column and start_row == end_row:
    #     return True
    #
    # elif moves == 0:
    #     return False
    #
    #
    # def movement()
    # return start_column

print(ex9("a1", "f5", 3))
print(ex9("a1", "f5", 2))